### PR TITLE
[MLIR][ONNX] Add support for onnx.scan op

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
This commit lowers onnx.scan op to torch.prim.Loop op and adds the lowering in the onnx pipeline.

### CODE REVIEW

**Comment by renxida** [comment]
Might be a good idea to either run an e2eshark op test or a custom test case with a contrived graph "loop body"

just to make sure this actually runs and is consistent with onnxruntime.

Sometimes the PrimLoopOp lowers to linalg but IREE is unable to execute it due to problems with mutability in tensors.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 1]
```diff
@@ -3931,4 +3931,113 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+          Value self = BlockArgument::Value(
+              primLoop.getRegion().getArgument(numInits + 1 + i));
+          auto scanOutSliceTy =
+              cast<Torch::ValueTensorType>(scanOutSlices[i].getType());
+          SmallVector<int64_t> unsqueezeShape(scanOutSliceTy.getSizes());
+          unsqueezeShape.insert(unsqueezeShape.begin(), 1);
+          Type unsqueezeTy = scanOutSliceTy.getWithSizesAndDtype(
+              unsqueezeShape, scanOutSliceTy.getOptionalDtype());
+          Value src = rewriter.create<Torch::AtenUnsqueezeOp>(
+              loc, unsqueezeTy, scanOutSlices[i], constZero);
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:4030]: You can use `Torch::unsqueezeTensor()` here.

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (Shukla-Gaurav) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Thanks :)

**Comment by Shukla-Gaurav** [comment]
> Might be a good idea to either run an e2eshark op test or a custom test case with a contrived graph "loop body"
> 
> just to make sure this actually runs and is consistent with onnxruntime.
> 
> Sometimes the PrimLoopOp lowers to linalg but IREE is unable to execute it due to problems with mutability in tensors.

Thanks @renxida for pointing out this. I am tracking the e2e test case here: https://github.com/nod-ai/SHARK-TestSuite/pull/276

**Comment by Shukla-Gaurav** [comment]
@vivekkhandelwal1 @renxida  can you please re-review and let me know if this looks good?

**Review by vivekkhandelwal1 — APPROVED** [review]

**Pull request merged**